### PR TITLE
Added broadcast check in scan.

### LINKED ISSUE #5384: Theano TypeError message far from useful
Currently I have encountered a TypeError in using T.grad to compute a gradient regarding outputs of a T.scan.
Most debugging methods provided by Theano involves with compiling the function, so I'm really out of choice right now. The error is:
TypeError: Cannot convert Type TensorType(float64, 3D) (of Variable Subtensor{:int64:}.0) into Type TensorType(float64, (False, True, False)). You can try to manually convert Subtensor{:int64:}.0 into a TensorType(float64, (False, True, False)).

Let's be serious, this kind of error message is utterly junk.  The suggestion _You can try to manually convert Subtensor{:int64:}.0 into a TensorType(float64, (False, True, False))_ is obviously useless, who knows where did that Subtensor comes from? 
It's not hard to catch this TypeError in theano calling stack(say, catch it in scan-op) and provide more useful messages. 
Theano should provide more information in exception handling.

### PR DESCRIPTION
**Pull request by Thrandis**
Adresses #5224 and fix #5384

I added an error message (it is still unclear for the moment), and a small test with the failure case.

### CODE REVIEW

**Comment by Thrandis** [comment]
Ok there is some failed tests I need to look at!

File: theano/scan_module/scan_op.py [hunk 1]
```diff
@@ -402,6 +402,32 @@ def make_node(self, *inputs):
                     'by using dimshuffle or shape_padleft. '
                     )
 
+        def _check_broadcast(v1, v2):
+            """ Checks that the broadcast pattern of v1 and v2.
```
Comment (nouiz) [inline theano/scan_module/scan_op.py]: remove "that"

File: theano/scan_module/scan_op.py [hunk 2]
```diff
@@ -402,6 +402,32 @@ def make_node(self, *inputs):
...
+
+            Controls that the broadcast pattern of the variable provided as
+            input to `scan` matches the broadcast pattern provided in
+            `output_info`. It raises an error when they don't match. The
+            typical case is when the user provides either the input or the
+            `output_info` (but not both) with a dimension fixed to 1,
+            which may wrongly be interpreted as broadcastable.
```
Comment (nouiz) [inline theano/scan_module/scan_op.py]: The method format() is call for other thing then output_info. I think this method _check_broadcast should not be called inside format(), but before the format() call for the handling of output info.

File: theano/scan_module/scan_op.py [hunk 3]
```diff
@@ -402,6 +402,32 @@ def make_node(self, *inputs):
...
+
+            """
+            msg = ("The broadcast pattern of the output of scan (%s) is "
+                   "inconsistent with the one provided in `output_info` "
+                   "(%s). The output on axis %d is `%r`, but it is `%r` on "
+                   "axis %d in `output_info`. This can happen if one of the "
+                   "dimension is fixed to 1 in the input, while it is still "
+                   "variable in the output, or vice-verca. You have to make "
+                   "them consistent.")
+            size = min(len(v1.broadcastable), len(v2.broadcastable))
```
Comment (nouiz) [inline theano/scan_module/scan_op.py]: Some of the errors are due to the fact that this check was done on other stuff then the output_info, including non_sequence that can be anything. This mean v1 and v2 do not always have broadcastable. I don't remember if moving the check to only output_info will fix the problem or not. In all cases, skiping the check if the broadcastable attribute isn't there would fix this.

Comment (Thrandis) [inline theano/scan_module/scan_op.py]: So I changed the check to be performed only if both variables have the `broadcastable` attribute, but I left it inside the `format` function.

File: theano/scan_module/scan_op.py (hunk 2 above)
Comment (Thrandis) [inline theano/scan_module/scan_op.py]: (see other comment)

File: theano/scan_module/scan_op.py (hunk 3 above)
Comment (nouiz) [inline theano/scan_module/scan_op.py]: Why you didn't move it? It could raise a bad error message in other cases.

**Comment by Thrandis** [comment]
I moved the check where it needed to be.

File: theano/scan_module/scan_op.py (hunk 3 above)
Comment (Thrandis) [inline theano/scan_module/scan_op.py]: Done!

**Comment by Thrandis** [comment]
I am wondering if my error message is clear enough! What do you think?

File: theano/scan_module/scan_op.py [hunk 4]
```diff
@@ -402,6 +402,35 @@ def make_node(self, *inputs):
...
+            if (not hasattr(v1, 'broadcastable') and
+                    not hasattr(v2, 'broadcastable')):
+                return
+            msg = ("The broadcast pattern of the output of scan (%s) is "
+                   "inconsistent with the one provided in `output_info` "
+                   "(%s). The output on axis %d is `%r`, but it is `%r` on "
+                   "axis %d in `output_info`. This can happen if one of the "
+                   "dimension is fixed to 1 in the input, while it is still "
+                   "variable in the output, or vice-verca. You have to make "
+                   "them consistent.")
```
Comment (nouiz) [inline theano/scan_module/scan_op.py]: Just add something like this to help guide the user on how to make them consistent:

    Using theano.tensor.{patternbroadcast,unbroadcast,addbroadcast} will help you do this.

**Pull request merged**